safenode/src/node/mod.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+mod api;
+mod error;
+mod event;
+
+pub use event::NodeEvent;
+use libp2p::request_response::ResponseChannel;
+
+use self::{
+    error::{Error, Result},
+    event::NodeEventsChannel,
+};
+use crate::{
+    network::{
+        Cmd, CmdResponse, Network, NetworkEvent, NetworkSwarmLoop, Query, QueryResponse, Request,
+        Response,
+    },
+    storage::{chunks::ChunkAddress, DataStorage},
+};
+use assert_fs::TempDir;
+use futures::StreamExt;
+use std::time::Duration;
+use tokio::task::spawn;
+
+/// Safe node
+#[derive(Clone)]
+pub struct Node {
+    network: Network,
+    storage: DataStorage,
+    node_events_channel: NodeEventsChannel,
+}
+
+impl Node {
+    /// Create and run the `Node`
+    pub async fn run() -> Result<(Self, NodeEventsChannel)> {
+        let (network, mut network_events, network_event_loop) = NetworkSwarmLoop::new()?;
+        let temp_dir =
+            TempDir::new().map_err(|_| Error::Other("failed to create temp_dir".to_string()))?;
+        let storage = DataStorage::new(&temp_dir);
+        let node_events_channel = NodeEventsChannel::default();
+        let node = Self {
+            network,
+            storage,
+            node_events_channel: node_events_channel.clone(),
+        };
+        let mut node_clone = node.clone();
+
+        // Spawn the network task for it to run in the background.
+        let _ = spawn(network_event_loop.run());
+
+        let _ = spawn(async move {
+            loop {
+                let event = match network_events.next().await {
```

Comment:
shall be the `try_next` call here?

Reply:
right. but be careful that `recv` call is a blocking call.
which means may result in a `dead lock`